### ISSUE TITLE
Add config option for MRU tabs

### ISSUE THREAD

**Issue by 50Wliu** [atom/atom#11035]
### Prerequisites
- [x] Can you reproduce the problem in [safe mode](https://atom.io/docs/latest/hacking-atom-debugging#check-if-the-problem-shows-up-in-safe-mode)?
- [x] Are you running the [latest version of Atom](https://atom.io/docs/latest/hacking-atom-debugging#update-to-the-latest-version)?
- [x] Did you check the [debugging guide](https://atom.io/docs/latest/hacking-atom-debugging)?
- [x] Did you check the [FAQs on Discuss](https://discuss.atom.io/c/faq)?
- [x] Are you reporting to the [correct repository](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages)?
- [x] Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3Aatom+-repo%3Aatom%2Felectron) to see if your bug or enhancement is already reported?
### Description
#10737 made ctrl+tabbing MRU-based, but did not offer a config option to revert to the existing left->right behavior.  Given that this is a very significant change to default behavior, we should keep the old sequential behavior as the default switching technique and expose MRU tabbing through its own setting.

/cc @natalieogle, @benogle
### Versions

You can get this information from executing `atom --version` and `apm --version` at the command line.

```
Atom    : 1.7.0-dev-6beed1d
Electron: 0.36.8
Chrome  : 47.0.2526.110
Node    : 5.1.1
```

**Comment by benogle** [comment]
I really think we should make MRU ordering the default. Many other editors ship with it as the default, many people want it out of the box, and there are already _3 other pairs_ of commands (`cmd-{|}`, `cmd-alt-left|right`, `ctrl-pageup|pagedown`) that move left / right through the tabs. 

If MRU switching is going to be the default, there will need to be a behavior change at some point. Why not now? I suppose the reason for why not now is if we're worried that it is buggy. But that's sort of what beta is for. If people are super worried about bugginess, maybe it wouldn't be the default for a release or so, then we'd need to do the behavior change.

But even then, I think it will be interesting to see if there are complaints or confusion when this hits beta. 

People can revert via a keymap change, which we can advertise in the release notes or the blog post:

```
'body':
  'ctrl-tab': 'pane:show-next-item'
  'ctrl-tab ^ctrl': 'abort!'
  'ctrl-shift-tab': 'pane:show-previous-item'
  'ctrl-shift-tab ^ctrl': 'abort!'
```

**Comment by mnquintana** [comment]
:+1:

**Comment by xak2000** [comment]
What this `^ctrl` and `abort!` really do and why simple rebinding `'ctrl-tab': 'pane:show-next-item'` is not enought? I think this syntax is not intuitive for many people. I thought long and hard what that might mean and why we need write `^ctrl` second time if we already wrote `ctrl-tab` and still I have no idea what this means. And I saw other people post similar feelings.

I think the problem for people who want back left-to-right behaviour not with defalut, but with this cryptographic settings. :)

**Comment by lee-dohm** [comment]
`abort!` is the same thing as `abortKeyBinding()` documented in the [Flight Manual](https://atom.io/docs/latest/behind-atom-keymaps-in-depth#overloading-key-bindings). The `^ctrl` is a new ability for the [atom-keymap package](https://github.com/atom/atom-keymap) that was added specifically for MRU tab switching, it means that the command isn't activated until <kbd>Ctrl</kbd> is released.

**Comment by xak2000** [comment]
Thanks for the answer. Now I read the docs and especially useful information in PR. It's really magic... :sparkles:

Now I understand what each of this construction do separately, but I still doesn't understand what they do in this concrete binding:

    'ctrl-tab': 'pane:show-next-item'
    'ctrl-tab ^ctrl': 'abort!'

My attempt to understand:
If user presses <kbd>Ctrl</kbd>+<kbd>TAB</kbd>, then do command `switch to next tab in left-to-right order`. If user then releases <kbd>Ctrl</kbd>, then abort this key binding (what keybinding? releasing <kbd>Ctrl</kbd> key? Or pressing <kbd>Ctrl</kbd>+<kbd>TAB</kbd>?) and continue searching for another matching binding. But why we need this? Maybe MRU switching in new release uses exactly this binding `'ctrl-tab ^ctrl'` and we just want to unbind it? But why we don't use `unset!` then? I'm totally confused now.

**Comment by lee-dohm** [comment]
> Maybe MRU switching in new release uses exactly this binding 'ctrl-tab ^ctrl' and we just want to unbind it?

:point_up: This is the reason. There's really very little difference between `abort!` and `unset!` in this instance. You can use either.

**Comment by demhydraz** [comment]
I think it'd still be useful to have a config setting for us international users - which either can't type properly, or apply a patch to disable keyup functionality in keymaps.

**Comment by UltCombo** [comment]
I'd say to keep the old behavior by default and add a config option for MRU tab switching.

Such changes to the defaults are very disruptive for users and, IMHO, this is a huge downgrade—sequential tab switching by default is one of the few things that Atom had got better than Sublime.

At least for me, MRU is very hard to work with once you are working with more than two tabs—I don't have enough spare cognitive load to remember and update the order of the tabs I've used recently, I'm spending it with things that actually matter such as code logic and flow. And in case I'm working with just a few tabs of a larger set, it is easy to reorder them to be next to each other and sequentially switch between them forwards/backwards, thus only a single tabs order needs to be remembered and you can always take a quick look at the tabs bar to refresh your memory. Therefore, sequential switching is superior in every single aspect and MRU is completely useless to me.

Okay, sorry for the small rant.

In any case though, I know I have no say here and you guys can do whatever you want, but please, please make sure to document a way to disable MRU when you announce v1.7.0 stable in the blog post announcement (and perhaps somewhere in the official documentation?) if it ships with MRU by default.

**Comment by jneubrand** [comment]
For anyone else annoyed by this change: according to my (limited) testing, `abort!` doesn't work with the config changes posted above. Use `unset!` instead :smile:

**Comment by BlaM** [comment]
So the fix is to insert this into the keymap, correct?

    'body':
      'ctrl-tab': 'pane:show-next-item'
      'ctrl-tab ^ctrl': 'unset!'
      'ctrl-shift-tab': 'pane:show-previous-item'
      'ctrl-shift-tab ^ctrl': 'unset!'

**Comment by Alhadis** [comment]
Yep, `unset!` appears to do the trick. Using `abort!` makes things even weirder. :|

In all honesty, this level of customisation shouldn't be necessary for something as fundamental as window navigation...

**Comment by perlun** [comment]
I agree that it wasn't perhaps so wise to let this slip into a "minor" release (1.6 -> 1.7) _without a UI_ to disable it. (I think I already had an MRU package in place, since I found the old behavior completely wrong for me, so I think the feature is nice, it's just the way that we delivered it to be a bit obtrusive.)

**Comment by Ajedi32** [comment]
Why not just make MRU the default for new installs, while keeping the old behavior _for existing installs_? Most users currently using Atom have either gotten used to the current tab switching behavior, or (like me) installed a package to switch to MRU tab switching a long time ago. Then for new installs make MRU the default with an option to disable it.

**Comment by jneubrand** [comment]
@Ajedi32 idk, IMHO I think it'd be fine if it's made _very_ clear in the release notes that there's a toggle (and where it is)

**Comment by BlaM** [comment]
How many people actually read the release nodes? Most users just want to have a stable development environment, only few actually have the time to actively manage it.

**Comment by Alhadis** [comment]
I read the release notes, but I fully agree. Most casual Atom users will see an upgrade, install it habitually in an age of evergreen browsers, and not think twice about what it added until they notice a new feature... or bug.

**Comment by phrz** [comment]
This is not at all intuitive — I have been looking over the course of days in Atom Discuss and these issues to find this exact issue. I use Xcode, Sublime, and Atom, and now Atom is the odd one out on tab behaviour — I have never touched an IDE with MRU tab ordering.

If Atom is going to implement a very opinionated change like this, there at least needs to be a Settings switch for it. 

Even if the prevailing opinion is that MRU is superior, the large group of more novice users who prefer the old tab behaviour have no switch, either in General Settings or the `tabs` package, to tweak this. I am not a beta tester, yet for the editor behaviour to change so radically and not provide me with an "out" makes me feel like one.

**Comment by UltCombo** [comment]
By the way, you guys say MRU was a "heavily requested feature", but I've read the [relevant thread](https://github.com/atom/atom/issues/5344) and I can see that most of Atom's core contributors opposed changing the default tabbing behavior to MRU. Also, you have to consider that people who were looking for MRU tabs switching in Atom are likely to reach that thread and +1 it, meanwhile casual Atom users who were happy with the sequential tab switching behavior and don't follow this repository (e.g. me) are very unlikely to reach such a thread. Therefore, your metrics were completely unfair and skewed!

The next time you consider such a drastic fundamental change to defaults, at the very least please perform a poll in an unbiased way for the community to decide (something simple like a Twitter poll may be enough, I think).

**Comment by maxnowack** [comment]
MRU is only intuitive if you can see the items ordered while switching.
Like `cmd-tab` on OSX or `alt-tab` on windows.

Please switch back to the old behavior or provide a config option for that! The workaround doesn't work for me ...

**Comment by Ajedi32** [comment]
@UltCombo Agreed. I think the same effect is also happening here. With the exception of one or two people, everyone in this thread got here by noticing the change, not liking it, then searching Google to find out what happened and how to fix it.

A poll of some kind would be a good idea here. Ideally we'd just have Atom ask users which option they prefer when they update, then use that selection to determine which option the editor users for them and which option becomes the default for new installs.

**Comment by AgentME** [comment]
It baffled me that the order that I cycle through tabs seemed to be random, and to add insult to injury, it would then change throughout the session. The logic of the ordering was entirely lost on me. I've always expected the tab behavior to be consistent with Chrome and had been very happy when I first tried Atom and found that it was. I typed up a bug report that tab order was completely broken by 1.7 and was about to report it before I found this thread.

**Comment by demhydraz** [comment]
I'd also like to add that the proposed fix doesn't work (on 1.8.0-beta1, Arch Linux x86_64)

**Comment by xak2000** [comment]
I just want to give my voice for MRU.

And yes, I read this thread only because I posted in it early. Other people who prefer MRU just never see this thread. Therefore discussions like this never be constructive. There will be always many people who don't like the subject and almost none who likes because they just don't ask google for it. They just say "Ok. I like this new feature." and that's it.

For existing users something like asking user when updating: "There is new MRU tab switching feature you can use. [Enable MRU tab switching] [Use current behaviour]" would be ideal. Of course only if current <kbd>CTRL-TAB</kbd> binding is default and not overriden.

For new installations... I don't know. But all (literally all!) editors which I used before were using MRU. And not only editors. Windows <kbd>ALT-TAB</kbd> always used it and I just got used to it since the days of Windows 95. I never used MacOS. Maybe MacOS uses left-to-right behaviour? I don't know. The only exception to this that I use is `Google Chrome` and me always enraged his left-to-right behavior. Especially when 20 tabs are opened and I only want to switch between two, but know that the other 18 will need soon, therefore not close them.

But the important thing with Windows and all other editors with MRU: If you do not release the modifier key after pressing <kbd>TAB</kbd>, then they show overlay window with current items placed in MRU order. And therefore you know the order.

MRU is very convinient when you need continuously switch between two items (tabs). But it can became very inconvinient if you really need to switch to **other** tab. And in this cases this overlay window with list of items is very important. Without it I can understand @AgentME feelings about randomness. :) I like how it's done in [tab-history](https://atom.io/packages/tab-history) extension. Used it before (when atom does not have this functionality) and it worked like a charm.

**Comment by jneubrand** [comment]
I think it's fine if there's no modal dialogue for users who upgraded. I just think it's sad that this wasn't mentioned in the release notes and that such a huge change to the default keybindings was made without keeping (simple) configurability in mind.

**Comment by alexrussell** [comment]
@xak2000 of course you're right about this issue unfairly representing the no-MRU camp, but, as was mentioned earlier by @UltCombo, the original thread that was apparently the basis for the feature's inclusion in Atom was the exact same but the other way around - all the MRU people posted there so the Atom core devs thought there was a lot of call for it and very few people against it. So yeah, creating a decision by asking the community will never be fair unless there's some way to ask every last user their preference.

That said, I say don't bother with the modal - it's too intrusive and more than a little odd to do. As has been said, most other editors use MRU by default, so it probably makes sense for Atom to follow anyway. But all of those editors give a simple setting to change it - not some crazy "if you don't want to opt in to this feature remap the keys and then unset the special keymaps put in place to make MRU actually work properly, it's as simple as these cryptic four lines, I can't believe I even had to spell it out for you!"

Also, FWIW @jneubrand it was mentioned in the release notes - `v1.7.0-beta0` and `1.7.0` stable both mention it. They don't go into any more detail but they do say that MRU is in.

**Comment by alfuken** [comment]
+1 for the option checkbox or the rollback. I've spent quite some time to at least figure out WTF, and how to fix this.

Right now both `pane:show-previous-item` and `pane:show-previous-recently-used-item` work absolutely the same, which is pure nonsense!

**Comment by mcmar** [comment]
It's totally insane that I opened my editor up after updating and my tab behavior was completely changed. This should absolutely have an option to disable it for the sanity of developers everywhere.

**Comment by Anon-Penguin** [comment]
@benogle

> 3 other pairs of commands (cmd-{|}, cmd-alt-left|right, ctrl-pageup|pagedown) that move left / right through the tabs.

FYI: None of these work on Windows for me with the CTRL equivalents except the page up/down, which is clunky on desktop and impossible on laptop. Furthermore, most users don't know this shortcut. There aren't viable alternatives for Windows users.

The keybinding did work, but it took me 15 minutes longer than it should've to find it (thanks to those who wrote it!).

**Comment by cryptoquick** [comment]
Boo, Atom. Boo. 😿

**Comment by valorin** [comment]
The suggested workaround keybinding **does not work** for me... tabs are still switching in some unknown and weird order. Please reconsider disabling MRU by default!

I'm on Ubuntu 16.06 x64 with:
```
> atom --version
Atom    : 1.7.2
Electron: 0.36.8
Chrome  : 47.0.2526.110
Node    : 5.1.1
```

**Comment by markidm** [comment]
I've been looking for a suitable, modern replacement as I've been using Notepad++ for longer than I'd care to admit, and I thought I'd finally found it. I was loving every feature of Atom for a good part of the day, but I was only working on two files. As soon as I added more files to the pane, my workflow fell apart as I kept alt-tabbing to the wrong file. This is seriously frustrating!

So, another +1 for the MRU switch option.

**Comment by Alhadis** [comment]
I'd also like to point out that Notepad++ has a MRU checkbox sitting in the very visible location of the first settings pane the user sees when they flip the options open.

Not having this behaviour controlled by a core preference is in every sense counter-intuitive.

**Comment by courthead** [comment]
Honestly spent the last week or two thinking this was a bug. Not a fan of this change at all.

**Comment by anomaly256** [comment]
Whether MRU  is default or not is really not the issue - the issue is that if a user wishes to change the behaviour back the process is convoluted and unintuitive.

I don't mind if MRU is made the default but there really should be a preferences toggle to revert to tab order if the user wishes it without having to dig through search results wondering why their keymap override is not having the desired effect only to discover 'ctrl-tab ^ctrl': 'unset!' is needed as well.

Honestly what does this string even mean?   '^ctrl' ?  'unset!' ?  I assume that exclamation mark has some critical relevance to this command which is not immediately clear but the caret+ctrl is even more confusing.  I could go digging through documentation or source code to discover this but I really shouldn't have to.  a) The UI claims keymaps.cson is an OVERRIDE which I believe means that needing the unset in the first place is nothing short of a regression/bug, and b) Can people really be expected to know that this specific cryptic string in your custom DSL is required without spending time examining the API involved??
&lt;/my_2_cents&gt;

**Comment by alexrussell** [comment]
@anomaly256 

I am 100% not defending this MRU thing as I agree with the general sentiment of this issue, but to vaguely address your questions/points, this keymap stuff is an unfortunate grey area:

The caret syntax was introduced specifically (as far as I can tell, at least) for this MRU feature, in a way that means that the MRU stuff doesn't have to use core hacks - Atom is supposed to be modular after all, and bits of code that get special treatment in the core is not ideal. It means "when this key is lifted, do the following command, but not until it is lifted". As such, 1. it is necessary in order to ensure that the "last tab you were on" is only remembered when you actually stop on it (imagine pressing and holding down ctrl and pressing tab a few times, you don't want each of those intermediate tabs to go into MRU list) and 2. your keymaps.cson still is actually an override - the problem is that you can override the ctrl+tab behaviour, and presumably you are, but the caret control keymap is a completely separate keymap command that needs overriding. Hence why you need to unset it - you are overriding it with an unset.

But yes, it's all very infuriating (and somewhat counterintuitive).

**Comment by courthead** [comment]
I agree with @anomaly256. It's simply not user friendly to release a change to default behaviour without (a) making it painfully obvious to those users that the change was made, and (b) explaining how to reverse it.

In this particular case, we should have gotten a simple popup after upgrading, similar to how Gmail, Twitter, etc display informational popups to explain new features they've added.

In addition, the path for reverting this change should've been more thoroughly tested. There's a bug where copying the new key binding from the key bindings table in the settings doesn't include the `^ctrl` part. Similarly, the key binding resolver also doesn't show the `^ctrl` part. Due to this, it's all but impossible to learn about the new `^ctrl` syntax and override the new keybinding, even for someone who knows and understands Atom's key binding system well. The fact that this bug exists means this change was made without really testing out the user experience of reverting it, which is a great way to frustrate a lot of people.

**Comment by anomaly256** [comment]
@alexrussell thanks for the explanation, and I can see how it's a bit of a pickle since there are actually 2 key bindings to override but only 1 is apparent/visible to the user.  I guess the 'copy' link on the keymapping page should account for releases as well and copy both actions that a key combination can result in.  I do think a preference pane addition to handle this 1 specific case is warranted though and easier on users.

@courthead I agree, the copying should include both press and lift actions a key combination can result in, then when a user sees them when they paste they can remove the ones they don't want to override.  Or at a minimum lift actions should be visible on the keymapping page directly underneath their press counterparts

**Comment by caseywatts** [comment]
The keybinding override doesn't work for me. While this issue is being sorted out, I've downgraded to Atom 1.6.2, before the breaking change. I just wanted to include this option in the thread in case it can help other people in the meantime:

- Install atom 1.6.2 https://github.com/atom/atom/releases/tag/v1.6.2
- Uncheck "Automatically Update" under core settings


(It seems that the "next tab" concept has MRU baked into it too? I'm not sure what's going on, but I don't want to spend time on this right now)

**Comment by dmo-odoo** [comment]
Same here @caseywatts, I'm unable to work with the current tab switching configuration. I could work out my mind to grasp which file I would go to when I ctrl+tab once, maybe twice, but how am I supposed to know on which file I'll end up if I press ctrl+tab four times ? How can I know how much time I have to ctrl+tab to reach that file over there ? I usually don't update often so I only noticed it recently. I'm very happy that I found this thread because I was going mad about this issue. Reverting back to Atom 1.6.2 it is then, yeah !

**Comment by apboon** [comment]
Was this argument already been raised?

Interface between human and computer should be as seamless as possible. Both human and computer should have the same information available / know the same things / talk about the same thing.

Left-to-right = predictable: A programmer can visually see how many taps on the keyboard's tab key is needed to navigate to a certain tab; tap = tab.
Metaphor/Analog: A model-M keyboard where keys stay in place enables you to type blindly.

MRU = unpredictable: A programmer might remember the last tab used (if accustomed to it) but other than that, most of us do not remember/follow the tab order past 2 or 3 tabs. Tabbing requires unreasonable effort (beyond 2/3 tabs) because the human does not know what the computer knows.
Metaphor/Analog: A model-M keyboard where keys constantly change places in order of word prediction to "help" the human type faster.

@xak2000: Windows `Alt` + `Tab` does not have MRU behavior (like Atom). Windows switches left-to-right but places the window previously used as the first next window, then continuing the other windows left-to-right. This is the most of "MRU" (variant) I have seen within other Windows programs that I use or have used (if I encountered such a program).

From Mozilla (Firefox) to Adobe (Reader/Photoshop) to Google (Chrome) to Microsoft (Visual Studio) to Corel (Paintshop) to Postman to WinSCP to ConTEXT.. they all use plain left-to-right.. (with `Shift` modifier to switch back)

Support for MRU, no matter its unpredictable nature, sure, I'll vote all for it! Everyone should feel comfortable and be able to customize his editor.
But lets make predictable tab-order (seamlessness between human and machine) the standard.

Alternatively, predictable can also be: left-to-right with "previous first next" like Windows, although Windows (`Alt` + `Tab`) is kind of a loner in its use.

On a different note: I have been using ConTEXT for years, trying out Sublime and the sorts, throwing them all away because they did not behave as expected (or actually because I could not find the right settings fast enough.. or at all..) Atom is the first editor that just did what it suppose to do out-of-the-box.. MRU by default might impact first impression more than you'd think..

You guys do what you want to do.. ultimately I care more about just having an on/off switch than how Atom is received by people about to make the switch to Atom for the first time.

**Comment by caseywatts** [comment]
aha! I accidentally investigated more :)

`cmd`+`.` shows the `keybinding resolver`, it shows how the keybinding is done. In my case, I'd `brew install`ed atom, and it comes with its own darwin keybinding. Somehow both its keybinding and my personal override one (that was recommended in this thread) were both being triggered.

![image](https://cloud.githubusercontent.com/assets/273653/15179929/9d0e74f8-174c-11e6-9150-f51f2980a97e.png)

I tried removing the 4 offending lines from the bundled keymap (by clicking on its filename in the bottom pane), but that didn't work (and it also doesn't seem to be the best way to fix this).

I'm trying this now, using a more specific css selector to add the command back, as sorta recommended [here](https://github.com/atom/atom-keymap/issues/110#issuecomment-216053644):
```
'body':
  'ctrl-tab': 'unset!'
  'ctrl-tab ^ctrl': 'unset!'
  'ctrl-shift-tab': 'unset!'
  'ctrl-shift-tab ^ctrl': 'unset!'

'.platform-darwin':
  'ctrl-tab': 'pane:show-next-item'
  'ctrl-shift-tab': 'pane:show-previous-item'
```

**Comment by xak2000** [comment]
@apboon Just to clarify: Windows <kbd>Alt</kbd> + <kbd>Tab</kbd> definitely **have** MRU behaviour.
Just now I have 5 windows open. Current active window (Chrome) is second on taskbar. I press <kbd>Alt</kbd> + <kbd>Tab</kbd> once, not releasing <kbd>Alt</kbd>. What I see in popup:
2(current) 1 4 5 3 (the index of window from taskbar). So it's a MRU, not just previous used inserted before others in left-to-right.

Not sure about "Visual Studio", not used it 10+ years. But "Visual Studio Code", which I use day-to-day uses MRU by default. But of course with popup window! So I must say it is popup window that missing. If it was present, most people would not even notice that Atom now uses MRU.

And you should understand that all this feelings about atom in comparison to other software that you used is **yours** personal feelings. For me, on the contrary, the lack of MRU was not good first impression. All editors that **I** used before were MRU. So it is just personal. So I aggree that it should be switch in config (and also popup window, because without it I better will use left-to-right than MRU).

**Comment by caseywatts** [comment]
maybe also some of my randomness was from this?
![image](https://cloud.githubusercontent.com/assets/273653/15180976/a09baf4a-1752-11e6-8721-42f9fd9f44e5.png)

**Comment by dmo-odoo** [comment]
@xak2000 While having such popup would help grasp the order in which the tabs are with respect to the ctrl+tab, I have always frowned upon them because why have another list of tabs in another order than the one you already have in the editor ? If the order of the tabs you have in the editor don't suit you, then reorder them. Or maybe make atom reorder the actual tabs instead? So MRU and left-to-right would actually do the same. It might look pretty strange however...

To be honest, I think the discussion for or against MRU is not really productive. I'm not sure I could convince any MRU user to user left-to-right nor would I be convinced by any argument to use MRU. I think the only possible exit is an easy option in Settings. I don't really care which one is the default behavior, you can't please everyone anyway, I just want an easy checkbox to choose the one I like.

**Comment by apboon** [comment]
@xak2000, this made me giggle as we are both kind of right; it gets more complicated even. You are right about MRU order of window/program placement in the `Alt` + `Tab` list. Tabbing through the list itself (while holding `Alt` + `Tab`) goes left-to-right though. This is kind of what I was getting at. You can predict that each time when you press `Tab`, you will jump to the next thumbnail (representing a window).
I compare the tabs in the top in the Atom editor with the `Alt` + `Tab` list you see during switching.

So.. yes, Windows does MRU by placing its "tabs" in MRU order into an overview/list, and then going through it left-to-right. You can see where you are navigating to; the fifth window is actually the fifth tab.

Atom places tabs in some initial order or the developer reorders them to get a better oversight, then Atom jumps through the tabs in seemingly random order.
If Atom would forcefully reorder its tabs by MRU order, and go through it left-to-right, the user would know exactly where he would end up. But I think forcefully reorder would not be a very good solution :p

I just now noticed the commonalities and difference between Windows and Atom. Windows also have "tabs", which is the taskbar. But Windows does not use the taskbar to highlight which window (in MRU order) it has currently selected (like Atom does now); because it would be confusing to go through a non-MRU list (taskbar/tabs) in MRU-order (or presenting any order and then going about it in another order).

Looking at Windows, in order to do MRU right and predictable, Atom needs a MRU ordered list to pop-up while `Ctrl` + `Tab`-bing.

MRU support yes. But needs an MRU ordered (pop-up) list to make sense. "Blind MRU" does not work in my opinion.

**Comment by apboon** [comment]
@dmo-odoo 
The discussion should indeed not be about MRU yes or no. Let everyone have the editor they love.
The discussion should be centered about these three flavors:

* Tab order (sequential)
* MRU order (pop-up; sequential)
* "Blind MRU" (unknown order)

Blind MRU being the current implementation.

**Comment by xak2000** [comment]
>Looking at Windows, in order to do MRU right and predictable, Atom needs a MRU ordered list to pop-up while Ctrl + Tab-bing.

@apboon  Exactly! This is what I'm talking about in almost every my post about atom MRU implementation (I usually call it "popup window").

And yes, of course selecting items in this popup window goes left-to-right (or rather, up-to-down). How else? :) The whole purpose to show this popup window is to let user **know** the current MRU order instead of **remember** or **predict** it. It is exactly how MRU works in Eclipse, Sublime, Intellij Idea, Webstorm, Phpstorm, Visual Studio Code etc.

About reordering tabs instead of show popup window.. I think this is too intrusive. Maybe it is just my habit. I never used any editor which reorder tabs in MRU order. Maybe it will be convinient. Maybe not. I think rather no than Yes. If almost all other editors that use MRU show popup instead of reordering tabs, I think there is a reason. I'm just a user. But I'm sure, when something big like Intellij IDEA developed, there are experts of UI design and usability that take into account all possibilities and choose the "least evil" of all.

**Comment by chamini2** [comment]
Another problem I see with this implementation (as stated in #11699) is that adding to the keybindings:

```
'body':
  'ctrl-tab': 'pane:show-next-item'
  'ctrl-shift-tab': 'pane:show-previous-item'
```

It should work as expected, without the `'ctrl-tab ^ctrl': 'unset!'` needed.

The [*method* called](https://github.com/atom/atom/blob/72721af3c63fd0d0bd0b0dd196ff28a3cab9ba11/src/pane.coffee#L347) when `ctrl` is released (`ctrl-tab ^ctrl`) should not affect the shown tab, it should just change the internal stack of tabs; which shouldn't make a difference if we don't use MRU switching. The method linked seems to not change the pane to show, making it very weird that it does change behaviour.

**Comment by dmo-odoo** [comment]
Well, is anyone against the following points ?
- keep the MRU feature but add a popup so you know which tab will come the next time you ctrl+tab
- add an option to choose whether you want to use MRU with the popup or classic left-to-right tabbing

If not, then the only discussion left is which one should be set as the default. And to be honest, I don't really care, I just want both above points... ;-)

**Comment by valorin** [comment]
MRU should be disabled by default, with classic left-to-right the default. 

Any user who wants MRU is likely to be an advanced user, who won't have any problems finding it in the settings to enable. Users who don't like MRU, or more importantly don't understand how it works, are more likely to give up and use something else - thinking it's a bug rather than a "feature".

**Comment by apboon** [comment]
* Keep the MRU feature but add a popup so you know which tab will come the next time you ctrl+tab
* Add an option to choose whether you want to use MRU with the popup or classic left-to-right tabbing
* MRU should be disabled by default, with classic left-to-right the default.

I second that!

**Comment by Ajedi32** [comment]
@valorin Atom's primary use case is as a development tool. Presumably the vast majority of it's users are "advanced users".

Anyway, [as I pointed out earlier](https://github.com/atom/atom/issues/11035#issuecomment-211474968), it would be wrong to assume that the prevailing opinion in this thread is a representative sample of all Atom's users. We can't really know what the default should be without doing a poll of a representative sample of Atom's users.

Also all of the points being discussed in these last few comments have been made many times already:

- "Display a pop-up of the MRU ordering" is covered by #11535
- "Make it optional" was the whole point of this very issue
- "Make left/right tab switching the default", has no dedicated issue, and so far nobody's really made a strong case for what the default behavior should be one way or the other. (And by a strong case I mean one backed up with actual data, not just "I like this way better".)

**Comment by phrz** [comment]
To assume that the average user is "advanced" is a very misguided approach to designing a widely accessible application. The left-right tab behaviour being the default is not an attempt to win a battle of opinion, but in fact undo the damage of the MRU switch, which itself was purely based on the preference of a few people.

**Comment by courthead** [comment]
Indeed, it seems to me the switch to MRU was done based on the opinions of a much smaller group of people than those who are against it.

Most importantly, MRU by default seems buggy to those who've never used it (I've been developing for 20 years and have never seen Atom's style of MRU), whereas left-right is intuitive to all people, even MRU users. Exactly zero issues were posted here mistaking left-right tabbing for a bug when it was the default. It's far more common in applications that support tabbing, e.g. every browser ever, the terminal, most IDEs, etc. Seems obvious to me what the default should be...

**Comment by Ajedi32** [comment]
> The left-right tab behaviour being the default is not an attempt to win a battle of opinion, but in fact undo the damage of the MRU switch

If the change was reverted days or hours after it was made perhaps, but I think that changing the default behavior _again_ at this point, months after the original change was made, would just create more confusion, if anything.

> Indeed, it seems to me the switch to MRU was done based on the opinions of a much smaller group of people than those who are against it.

That's certainly possible, but again, [we don't know that](https://github.com/atom/atom/issues/11035#issuecomment-211474968) without actually measuring it.

> Most importantly, MRU by default seems buggy to those who've never used it

I think that's more of a side effect of the current implementation not displaying a pop up of the MRU stack (like literally every other MRU switching implementation I've used) than an inherent aspect of MRU.

**Comment by valorin** [comment]
> If the change was reverted days or hours after it was made perhaps, but I think that changing the default behavior again at this point, months after the original change was made, would just create more confusion, if anything.

The users who don't like MRU would see it as an **overdue bugfix**, and those who do... well... that's what release notes are for! (Yes, that's the same miss-guided argument used *for* the switch to default... not such a good idea now, hey?)

**Comment by courthead** [comment]
> If the change was reverted days or hours after it was made perhaps, but I think that changing the default behavior again at this point, months after the original change was made, would just create more confusion, if anything.

I disagree. The confusion is primarily caused by the poor user experience of the MRU feature itself, less so by changing between systems to the next. For people (like me) who endured MRU for weeks thinking it was a bug before finding this thread, this would be seen as a bug fix. And for people who like MRU, it would seem like a reversion to a tabbing system with a non-broken UX, since the current MRU UX is basically unusable.

Even after the MRU UX is fixed (via a popup), I still think there's a stronger case to make left-right tabbing the default, and make MRU opt-in. It's just a less common tabbing system in general, and although we may never be able to take an actual vote on what people want, the limited evidence we do have to go on suggests that people like left-right tabbing...

**Comment by xak2000** [comment]
@courthead We don't have any evidence because, as @Ajedi32 said several times, this thread is almost exclusively filled with people who don't like MRU. Those who like it just don't google it.

Also we must take into account that it's not MRU is bad. It's implementation without popup is bad.

> It's just a less common tabbing system in general

Completely disagree. Among programmers, MRU is more common tabbing system. Don't know which code editors you used before, but Chrome is **not** code editor.

But I'm not in any way not saying you are wrong and I am right. I'm just saying that everyone has their own opinion and their own experience. You should not take **your own** experience for the fact (so am I).

I'm just can't confortably switch between tabs without MRU. As much as you can't do without left-to-right. How many atom users are like me and like you? We *don't know* because this thread can't be taken as measure.

Anyway, many times it was suggested that there should be just 2 options and switch. Which option would be default? I don't care, really. I think MRU should be default (good MRU, with popup), because it is default in others cool code editors. But it just me. If left-to-right would be default, it's ok.

So let's close the discussion about which is better. There is no better. There is better for one group of people and better for others.

I think there is so many posts in this thread, that the developers of Atom is unlikely to be able to grasp the essence:

- Current MRU implementation is broken (should be popup).
- Current enable/disable of MRU is hard (should be simple switch).

**Comment by perlun** [comment]
I completely agree with @xak2000. And, I definitely was bugged by the previous defaults, and find the current MRU UX not "unusable", instead it works quite well _for me_, expecting MRU and finding anything else weird...

Could someone just add an option for disabling MRU and file a PR about it? (If it hasn't already been done) That should be the "quick fix" that makes things loads better than they currently are.

**Comment by courthead** [comment]
> @courthead We don't have any evidence because, as @Ajedi32 said several times, this thread is almost exclusively filled with people who don't like MRU. Those who like it just don't google it.

There are other threads, too. It seems to me the number of people complaining about MRU now that it's the default is much higher than the number of people who complained about LTR while it was the default.

But I digress. I think we mostly all agree on the important issues: MRU will be much more usable with a popup, and there should be a simple config option to switch between the two systems.

**Comment by Alhadis** [comment]
> It's like if I upgraded and you changed my font to something completely different.

OH GOD, DON'T EVEN PUT THAT IMAGE INTO THEIR HEADS. PLEASE.

**Comment by emlazzarin** [comment]
None of the ludicrously complex workarounds have solved this bug for me (OS X; Atom 1.7.3).

**Comment by 50Wliu** [comment]
> What, specifically, is being done to fix this and what is the recommended workaround?

There is currently no work being done to "fix" this.  As @lee-dohm described in https://github.com/atom/atom/issues/5344#issuecomment-215899975, the best way to approach this would be to move discussion to Discuss (heh) until a concrete proposal materializes.  The recommended workaround is to follow the instructions in [this comment](https://github.com/atom/atom/issues/11035#issuecomment-192015334) if you want to revert to left->right tabbing.

**Comment by courthead** [comment]
@50Wliu What do we need to do to make a concrete proposal materialize? At the very least, it looks like adding a simple config option for reverting the MRU change is universally supported.

**Comment by iolsen** [comment]
Core team manager here. We hear you and definitely want to fix this. We will eventually if the community doesn't beat us to it. It seems the consensus is that we need:

1. An option to enable/disable MRU that modifies the keybindings.
2. A UI showing the traversal order when moving through tabs in MRU order, e.g. like [tab-switcher](https://atom.io/packages/tab-switcher).

We'd readily accept a PR to the effect.

**Comment by alfuken** [comment]
@iolsen why not just extract the MRU feature along with UI to a separate core package? Those who like it, will keep it. Those who don't will just disable the package.

**Comment by Alhadis** [comment]
Don't you think it's overkill to have two tab-specific core packages?

**Comment by alfuken** [comment]
@Alhadis still better than having a fat core app. There are now about ~80 core packages. And I, for example, have disabled 30 of them.
And it's good.
Otherwise Atom will turn into Yet-Another-Monstrous-IDE-That-Can-Brew-You-A-Coffee. So, the more disable-able features, the better.

**Comment by Alhadis** [comment]
I'd kill to have Atom brew me a coffee, actually.

Would complement its CoffeeScript-based API nicely.

**Comment by alfuken** [comment]
@Alhadis then write a package for that! :) Just make sure your coffee machine have an API ;) &lt;/offtopic>

**Comment by alfuken** [comment]
Still open... huh. Why not assign it to @natalieogle ? ;)

**Comment by Keavon** [comment]
I used to use Ctrl Tab and Ctrl Shift Tab all the time, but since this update the shortcut has been rendered entirely unusable because of its nonstraightforward and unpredictable behavior. It is, in my opinion, counter to human-centric design and should not be the default. Major ubiquitous applications like Chrome and Firefox go left right, and this MRU behavior differs from that common standard. Perhaps some other IDEs do the same, but they are less common than web browsers and they can't be assumed to use the better behavior. Please make left/right tabbing the default and add an option to set it to MRU. Or at the very least, at least add the option to tab left/right so I can even use the keyboard shortcuts again.

**Comment by sangeeth96** [comment]
With tab-switcher, MRU is pretty much usable but otherwise would be very difficult since I can't predict which tab is going to be selected next. Sometimes, I'd have to riddle through many tabs just to get to the tab very next to the currently active one. Hope we get an option to switch the behavior pretty soon.

**Comment by lee-dohm** [comment]
@Keavon There are instructions in the [Atom FAQ](https://discuss.atom.io/c/faq) on how to restore in-order tab traversal:

https://discuss.atom.io/t/i-want-ctrl-tab-to-work-like-it-used-to-how-do-i-do-that/28639/1

This Issue is only for adding a checkbox that does what the FAQ does :point_up:

**Comment by Keavon** [comment]
@lee-dohm Fantastic, thanks so much for those instructions! It still should be the default and there still should be a checkbox, but at least I can work productively in the meantime.

**Comment by dmo-odoo** [comment]
http://www.webupd8.org/2016/08/atom-190-released-with-drag-and-drop.html

1.9 has been released and, as far as I know, there is still no option to disable MRU tab switching.

**Comment by iolsen** [comment]
Related: atom/tabs#314, atom/tabs#361

**Comment by entozoon** [comment]
I whipped up a simple package to disable MRU tabbing (using @BlaM's keymap suggestion).
Search for `disable-mru-tabbing` in settings > packages to install it.
https://github.com/entozoon/atom-disable-mru-tabbing

**Comment by iolsen** [comment]
Noting some work in progress on this:

>1. An option to enable/disable MRU that modifies the keybindings.

Still TODO.

>2. A UI showing the traversal order when moving through tabs in MRU order, e.g. like [tab-switcher](https://atom.io/packages/tab-switcher).

- [ ] Catch keyups more effectively: https://github.com/atom/atom-keymap/pull/156
- [ ] New pane switch events to enable MRU List UI: https://github.com/atom/atom/pull/13008
- [ ] MRU List UI: https://github.com/atom/tabs/pull/388

**Comment by wyqydsyq** [comment]
It boggles me as to why when it became apparent making MRU the default method of tab switching without a UI to show the MRU list and without a UI option to disable it was a bad idea that someone didn't revert that.

I just did a fresh Atom install 5 months after I first had this issue and the fact that I still had to search Google to copy-paste some lines into my `keybindings.json` just to have my tabs behave predictably is absurd.

Based on the current state of this ( @iolsen says the MRU List UI is still TODO ) MRU should not be the default, it should disabled by default and with an *opt-in* checkbox on the Core settings page.

I understand it was decided to use MRU as the default to appeal more to ST users since Atom is more or less an open-source ST built with Electron, but when the issue was raised that UI-less MRU behaviour is confusing and unintuitive the change to make it default really should have been reverted.

**Comment by Keavon** [comment]
Indeed, imagine if Chrome suddenly switched its tab behavior to MRU. People would freak out and there would probably be a mass exodus from the browser because a logical feature was suddenly given illogical, unexpected behavior.

**Comment by iolsen** [comment]
Definitely some mistakes were made pushing this out too fast. 😞

**Comment by wyqydsyq** [comment]
@iolsen it's not really a huge deal, I was just surprised to find the situation with this hadn't changed in 5 months given how rapidly Atom had been moving in other areas. I guess that's the thing with OSS though, often only the fun sexy parts get attention, so thanks and good on you for being the one to finally step up on this issue 👍

**Comment by TheNotary** [comment]
Tabbing behavior... changed... in my latest... Atom update...

![http _mashable com_wp-content_uploads_2013_07_star-trek](https://cloud.githubusercontent.com/assets/799247/20014765/5bd134ba-a286-11e6-82f6-118d8b4fe7b1.gif)

**Comment by entozoon** [comment]
It hasn't has it?

**Comment by Ajedi32** [comment]
Looks like this bug was mistakenly closed by atom/atom-keymap#156. GitHub apparently isn't smart enough to realize that

> and is a prerequisite for adding the MRU list UI (atom/tabs#388) that fixes atom/atom#11035.

isn't the same thing as

> fixes atom/atom#11035

:stuck_out_tongue_winking_eye:

**Comment by pehrlich** [comment]
Is this still a current fix? With Atom `1.12.2` on OSX, the above code snippet does not revert the frustrating MRU default :/

```
'body':
  'ctrl-tab': 'pane:show-next-item'
  'ctrl-tab ^ctrl': 'unset!'
  'ctrl-shift-tab': 'pane:show-previous-item'
  'ctrl-shift-tab ^ctrl': 'unset!'
```

**Comment by iolsen** [comment]
> Is this still a current fix?

Yes it's still current in 1.12.x: I just confirmed it in 1.12.6. You might be able to use the keybinding resolver (cmd-.) to determine what's happening in your install.

**Comment by entozoon** [comment]
@pehrlich looks right, the plugin:
https://github.com/entozoon/atom-disable-mru-tabbing
Still works for me, on Atom and Atom Beta.

**Comment by iolsen** [comment]
Fixed in https://github.com/atom/tabs/pull/388. Set to ship in 1.14.

**Comment by Keavon** [comment]
Thank you!

**Comment by dmo-odoo** [comment]
Thank you so much !

**Comment by alexrussell** [comment]
@iolsen, a god amongst men.

**Comment by dpanter** [comment]
@entozoon Thank you! A mystery why the MRU switching crap isn't easy to disable, or even why it isn't disabled by default.

**Comment by iolsen** [comment]
@dpanter the option to disable it shipped 9 months ago in Atom 1.14.

**Comment by dpanter** [comment]
Finding that option was really easy too! It was on display in the bottom of a locked filing cabinet stuck in a disused lavatory with a sign on the door saying ‘Beware of the Leopard'.

Jokes aside, here is the built-in option: https://discuss.atom.io/t/i-want-ctrl-tab-to-work-like-it-used-to-how-do-i-do-that/28639

**Comment by iolsen** [comment]
😆 that's fair. We've been meaning to add setting search for a long time.

**Comment by dkurzaj** [comment]
Is there any actual tick-box to enable or disable MRU, or is the official solution to edit the keymap.cson file ?

**Comment by Ben3eeE** [comment]
@dkurzaj there is a setting in the tabs package see https://discuss.atom.io/t/i-want-ctrl-tab-to-work-like-it-used-to-how-do-i-do-that/28639
